MdeModulePkg/Variable: Move RT cache buffer allocation to DXE

makubacki opened the pull request:
## Description

Commit `d8f513de3e3ef228af7e6facf0ad3e35c3224032` (an edk2 commit) added runtime memory allocation to VariablePei. That will cause runtime memory bucket fragmentation affecting hibernate stability as these runtime allocations are separate from the main runtime memory type bucket allocation made by the DXE Core.

To preserve the existing functionality but allow the runtime buffer to come from the DXE bucket, the allocations in PEI are changed to boot services data allocations and the contents are moved to the runtime memory buffers (and those pages unblocked to MM) in the DXE variable driver after the bucket has been allocated by the DXE core.

---

> Note: I will upstream this to edk2 but it is urgently needed in Project Mu.

- [x] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?

## How This Was Tested

- CI
- Boot to EFI shell on QEMU Q35 and verify allocations come from the DXE allocated bucket buffer

## Integration Instructions

- N/A

makubacki committed:
MdeModulePkg/Variable: Move RT cache buffer allocation to DXE

Commit `d8f513de3e3ef228af7e6facf0ad3e35c3224032` added runtime
memory allocation to VariablePei. That will cause runtime memory
bucket fragmentation affecting hibernate stability as these runtime
allocations are separate from the main runtime memory type
bucket allocation made by the DXE Core.

To preserve the existing functionality but allow the runtime buffer
to come from the DXE bucket, the allocations in PEI are changed to
boot services data allocations and the contents are moved to the
runtime memory buffers (and those pages unblocked to MM) in the
DXE variable driver after the bucket has been allocated by the DXE
core.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki commented:
I was originally planning to take this straight to edk2, but it's needed here now. I'll add `MU_CHANGE` tags before completing it.

makubacki commented:
> I was originally planning to take this straight to edk2, but it's needed here now. I'll add `MU_CHANGE` tags before completing it.

Done - `MU_CHANGE` tags added

os-d commented on MdeModulePkg/Universal/Variable/RuntimeDxe/VariableSmmRuntimeDxe.c:
```diff
@@ -1577,9 +1586,123 @@ InitVariableCache (
+
+        Status = MmUnblockMemoryRequest (RuntimeBufferPhysicalAddress, Pages);
+        if ((Status != EFI_UNSUPPORTED) && EFI_ERROR (Status)) {
+          return Status;
```
If I'm reading this correctly, in the original implementation if we failed here, the mVariableRtCacheInfo.*Buffer pointer would not be set (either null or worse garbage). In this case, it will point back to the boot services buffer. Would it be safer to null the pointers in cases of failures (here and all the other error cases) so that runtime service code doesn't try to access boot services data? This is just an optimization, right, so we could meaningfully make progress if this was cleaned up?

If we do, then we should either continue and try to cleanup the other buffers, or null all buffers (I don't know if they have to be used as a group or not).

makubacki replied:
Addressed in https://github.com/microsoft/mu_basecore/pull/1519/commits/a88d5062cab74e247ab4af95e356773a2d290b9f

makubacki committed:
MdeModulePkg/VariableSmmRuntimeDxe: Disable var RT cache on alloc failure

If a RT data buffer fails to be allocated or unblocked, clean up the
RT cache state by (1) freeing any allocated buffers and (2) resetting
the buffer addresses back to zero to indicate the RT cache is not
available.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
